Enable setting VNet peering properties

nprokopic opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

/kind feature

**What this PR does / why we need it**:

This PR enables setting additional properties for VNet peering (`AllowForwardedTraffic`, `AllowGatewayTransit`, `AllowVirtualNetworkAccess` and `UseRemoteGateways`).

One use case for setting these properties to non-default values is to, for example, [configure VPN gateway transit for virtual network peering](https://learn.microsoft.com/en-us/azure/vpn-gateway/vpn-gateway-peering-gateway-transit).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3183

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Enable setting VNet peering properties (`AllowForwardedTraffic`, `AllowGatewayTransit`, `AllowVirtualNetworkAccess` and `UseRemoteGateways`)
```

willie-yao reviewed (commented):
This looks great, thanks for working on this @nprokopic !

Just one comment from my end: would you be able to add a unit test for `VnetPeeringSpecs()` in `cluster_test.go`?

Jont828 commented on api/v1beta1/types.go:
```diff
@@ -124,6 +124,42 @@ type VnetPeeringClassSpec struct {
+	AllowVirtualNetworkAccess *bool `json:"allowVirtualNetworkAccess,omitempty"`
+
+	// UseRemoteGateways specifies if remote gateways can be used on this virtual network.
+	//
```
Can probably remove this extra line

Jont828 commented on api/v1beta1/types.go:
```diff
@@ -124,6 +124,42 @@ type VnetPeeringClassSpec struct {
+
+	// UseRemoteGateways specifies if remote gateways can be used on this virtual network.
+	//
+	// If the flag is set to true, and allowGatewayTransit on remote peering is also true, virtual network will use
```
```suggestion
	// If the flag is set to true, and allowGatewayTransit on remote peering is also set to true, the virtual network will use
```

Jont828 commented on api/v1beta1/types.go:
```diff
@@ -124,6 +124,42 @@ type VnetPeeringClassSpec struct {
+	// UseRemoteGateways specifies if remote gateways can be used on this virtual network.
+	//
+	// If the flag is set to true, and allowGatewayTransit on remote peering is also true, virtual network will use
+	// gateways of remote virtual network for transit. Only one peering can have this flag set to true. This flag cannot
```
```suggestion
	// the gateways of the remote virtual network for transit. Only one peering can have this flag set to true. This flag cannot
```

Jont828 commented on azure/services/vnetpeerings/vnetpeerings_test.go:
```diff
@@ -65,6 +66,30 @@ var (
+		RemoteVnetName:            "spoke-vnet",
+		RemoteResourceGroup:       "spoke-group",
+		SubscriptionID:            "sub1",
+		AllowForwardedTraffic:     to.BoolPtr(true),
```
```suggestion
		AllowForwardedTraffic:    pointer.Bool(true),
```
I think we want to switch to the `"k8s.io/utils/pointer"` library

Jont828 reviewed (commented):
Thanks for the contribution! As @willie-yao said, we might want to add some tests in cluster_tests.go, and here are a couple points from me.

nprokopic commented:
@Jont828 @willie-yao I have added missing tests and also addressed other suggestions from the review. All tests are :green_circle: :)

Can we add this PR to the `v1.9` milestone?

Jont828 commented:
/lgtm

This looks great! Unless there are any objections, I don't see why not.

willie-yao commented:
/milestone v1.9

nawazkh commented:
Sweet! Thanks for adding this! 🚀

nprokopic commented:
> @willie-yao: You must be a member of the [kubernetes-sigs/cluster-api-provider-azure-maintainers](https://github.com/orgs/kubernetes-sigs/teams/cluster-api-provider-azure-maintainers/members) GitHub team to set the milestone. If you believe you should be able to issue the /milestone command, please contact your Cluster API Provider Azure Maintainers and have them propose you as an additional delegate for this responsibility.

Can somebody from @kubernetes-sigs/cluster-api-provider-azure-maintainers help out here?

mboersma commented:
/milestone v1.9

CecileRobertMichon commented:
/approve

Thanks all for the reviews 🚀
